### DISCUSSION TITLE
How to add my learned detection model

### DISCUSSION THREAD

**Discussion by yerin723** [mindee/doctr#1303] [General]
hello.
A new dataset was learned for drawing PDF OCR.
All you need are the numbers in the drawing PDF.
However, due to poor numerical detection ability, we learned a new dataset, and we currently have about 66 PDF files, of which there are a total of 2 detected classes, and the number of labels is about 160 per PDF.

https://github.com/mindee/doctr/tree/main/references/detection
I created a data format by referring to the link above and am currently learning the model.
Labeling was done through labelImg.

!python3 /home/piai/Desktop/ocr/doctr/references/detection/train_pytorch.py ​​/home/piai/Desktop/ocr/test /home/piai/Desktop/ocr/val db_resnet50 --epochs 10 --lr 0.01 - -b 4 --device 0

I ran it like this

Validation set loaded in 0.0743s (17 samples in 5 batches)
Train set loaded in 0.1859s (66 samples in 16 batches)
Validation loss decreased inf --> 63916.6: saving state...
Epoch 1/10 - Validation loss: 63916.6 (Undefined metric value, caused by empty GTs or predictions)
Validation loss decreased 63916.6 --> 6.42796: saving state...
Epoch 2/10 - Validation loss: 6.42796 (Recall: 0.00% | Precision: 0.00% | Mean IoU: 0.61%)
Validation loss decreased 6.42796 --> 4.2786: saving state...
Epoch 3/10 - Validation loss: 4.2786 (Recall: 0.00% | Precision: 0.00% | Mean IoU: 0.01%)
Validation loss decreased 4.2786 --> 3.99184: saving state...
Epoch 4/10 - Validation loss: 3.99184 (Recall: 0.00% | Precision: 0.00% | Mean IoU: 0.02%)

As such, the results are not very good.

1. How can I increase detection precision?
2. The resolution of the current image is too large, so if you increase the batch size, the GPU will explode.
3. I would like to test the detection after the model weight file (.pt) is created. Please kindly tell me how.
4. I plan to use db_resnet50.

Thank you!!

**Comment by felixT2K** [comment]
Hi @yerin723 👋🏼,

First you should finetune the model and not train it from scratch by passing `--pretrained` to the training cmd.
How to load custom models: https://mindee.github.io/doctr/latest/using_doctr/custom_models_training.html
Maybe @aminemindee can give more information because i don't think that the 66 files are enough also for finetuning

**Reply by yerin723** [reply]
db_resnet50 --epochs 5 --b 4 --lr 0.00001 --pretrained --device 0 
So I trained the model, but
Recall, Precision, and Mean IoU respectively
The results were 0.28%, 0.17%, and 2.52%.

Additionally, I have another question.
I know that when pretrained, an existing model is imported.
How many classes does the existing model have? I additionally learned data with two classes. Are there any problems?

**Reply by felixT2K** [reply]
The pretrained models are only trained on 1 class (text) for pure text detection.
docTR provides no pretrained KIE models currently only the text detection models which can be used for fine tuning on KIE.
There should be no problem but as mentioned i think 66 samples are much to less.
The models are pretrained on ~100K labeled document type samples.

**Reply by yerin723** [reply]
Still, are there things that can be fine-tuned, such as appropriate epochs, batch size, learning rate, etc.?

**Reply by felixT2K** [reply]
batch size depends on your hardware (put much as possible before OOM)
Your actual learning rate is "super save" but costs time i think 0.0001 works also the same way for fine tuning
epochs use more ~100 and stop the training if loss stops to decrease over 2-3 epochs (We don't have an EarlyStopping Callback in the training script currently)

But at the end to get better results the only solution i see is to increase your dataset.

**Reply by yerin723** [reply]
First of all, thank you! If you have any further questions, please leave a question!

**Reply by yerin723** [reply]
https://github.com/clovaai/synthtiger
 I understand that SynthTIGER is a method of generating training data for recognition.
So, is there another way to generate training data for detection?
This is because the dataset I have is too small.
I want to OCR number detection in a PDF document.

**Reply by yerin723** [reply]
I want to detect numbers in scanned documents rather than scean characters.
Since the dataset I have is too small, I don't think I need to load all the existing datasets.

So, I want to import only document-specific datasets and combine them with my dataset to see the results.
Tell me how to do it
